va/validation-authority.go

```diff
@@ -96,21 +96,27 @@ func (va ValidationAuthorityImpl) validateSimpleHTTP(identifier core.AcmeIdentif
 	}
 	hostName := identifier.Value
 
+	addr := va.getFirstAddr(identifier.Value, &challenge)
```

Comment:
This is a weird pattern, to pass in the challenge to be modified. I think it would be more idiomatic to return an error. But we want getFirstAddr to be able to do the work of constructing a core.ProblemDetails. Conveniently, ProblemDetails implements the Error interface, so we can have getFirstAddr return the appropriate ProblemDetails as an error. Then here in validateSimpleHTTP we would check for error, assign the ProblemDetails into the challenge, and set status = Invalid.

One problem with this is that setChallengeErrorFromDNSError follows the modifying pattern. I'd recommend also changing that method to return an error (i.e., turn a DNS error into a ProblemDetail). You'd presumably also rename it appropriately.